### ISSUE TITLE
The computational graph might be broken.

### ISSUE THREAD

**Issue by YaoxinHuang** [Saumya-Gupta-26/topoloss-cr#1]
You have used the numpy package to the result of Network, then the computational graph might be broken, which means you cannot autograd and back propagation accurately. 
Pls let me know how are you going to solve this problem?

**Comment by Saumya-Gupta-26** [comment]
Hello, it depends on your loss function. Theoretically, is the loss function differentiable/backprop or not? We use non-differentiable [TDA analysis](https://github.com/Saumya-Gupta-26/topoloss-cr/blob/983630a1a25eb5f239bb63a829b9642e394ed09e/topoloss_saum.py#L32) to determine which pixels in the prediction to penalize. Then we use [differentiable MSE](https://github.com/Saumya-Gupta-26/topoloss-cr/blob/983630a1a25eb5f239bb63a829b9642e394ed09e/topoloss_saum.py#L258) loss to penalize these pixels. Hence the computation graph is not broken.

Additionally, you can comment out the gudhi import since we use cripser library in the code

**Comment by YaoxinHuang** [comment]
> Hello, it depends on your loss function. Theoretically, is the loss function differentiable/backprop or not? We use non-differentiable [TDA analysis](https://github.com/Saumya-Gupta-26/topoloss-cr/blob/983630a1a25eb5f239bb63a829b9642e394ed09e/topoloss_saum.py#L32) to determine which pixels in the prediction to penalize. Then we use [differentiable MSE](https://github.com/Saumya-Gupta-26/topoloss-cr/blob/983630a1a25eb5f239bb63a829b9642e394ed09e/topoloss_saum.py#L258) loss to penalize these pixels. Hence the computation graph is not broken.
> 
> Additionally, you can comment out the gudhi import since we use cripser library in the code


Really appreciate ur patience, it helps a lot.

But there's also something disturbing me. You adopted the ```MSE Loss``` to maintain the continuity of the computational graph. However, the ```MSE Loss``` is the function of ```TDA_IMAGE```, which means the image after TDA, and I suppose that the so-called ```TDA_IMAGE``` is not a tensor, without grad, 'cause the ```gudhi``` package is built for ```numpy.ndarray```.  Then the chain of grad is not supported by the ```TDA_IMAGE``` to back-forward for those network parameters.

Thanks very much.

**Comment by Saumya-Gupta-26** [comment]
Hi, please read [this paper](https://proceedings.neurips.cc/paper/2019/file/2d95666e2649fcfc6e3af75e09f5adb9-Paper.pdf) to understand why the loss is differentiable and the computational graph doesn't break. Additionally, the original repo for the paper is [here](https://github.com/HuXiaoling/TopoLoss). You can contact the authors for more information.
